control-plane/agents/core/src/core/specs.rs

```diff
@@ -35,6 +37,416 @@ enum SpecError {
+            Ok(val) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.commit_op();
+                let deleted = registry.delete_kv(&key.key()).await;
+                match deleted {
+                    Ok(_) => {
+                        Self::remove_spec(locked_spec, registry).await;
+                        let mut spec = locked_spec.lock().await;
+                        spec.commit_op();
+                        Ok(val)
+                    }
+                    Err(error) => {
+                        let mut spec = locked_spec.lock().await;
+                        spec.set_op_result(true);
+                        Err(error)
+                    }
+                }
+            }
+            Err(error) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.clear_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.clear_op();
+                        Err(error)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(false);
+                        Err(error)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Start a destroy operation and attempt to log the transaction to the store.
+    /// In case of error, the log is undone and an error is returned.
+    async fn start_update(
+        registry: &Registry,
+        locked_spec: &Arc<Mutex<Self>>,
+        status: &Self::Status,
+        update_operation: Self::UpdateOp,
+    ) -> Result<Self, SvcError>
+    where
+        Self: PartialEq<Self::Status>,
+        Self: SpecTransaction<Self::UpdateOp>,
+        Self: StorableObject,
+    {
+        let mut spec = locked_spec.lock().await;
+        let spec_clone = spec.start_update_inner(status, update_operation, false)?;
+        drop(spec);
+
+        if let Err(error) = registry.store_obj(&spec_clone).await {
+            let mut spec = locked_spec.lock().await;
+            spec.clear_op();
+            Err(error)
+        } else {
+            Ok(spec_clone)
+        }
+    }
+
+    /// Checks that the object ready to accept a new update operation
+    fn start_update_inner(
+        &mut self,
+        status: &Self::Status,
+        operation: Self::UpdateOp,
+        reconciling: bool,
+    ) -> Result<Self, SvcError>
+    where
+        Self: PartialEq<Self::Status>,
+    {
+        // we're busy right now, try again later
+        let _ = self.busy()?;
+
+        if !self.state().created() {
+            Err(SvcError::PendingCreation {
+                id: self.uuid(),
+                kind: self.kind(),
+            })
+        } else {
+            // if it's not part of a reconcile effort
+            // validate the operation itself against the spec and status
```

Comment:
It seems as though we are only validating against the status.

Reply:
we're validating the spec against the status?

Reply:
I've improved it a bit by making the spec/status check explicit. For now it's still just a wrapper around the PartialEq but I think we'll probably move that directly to the impl instead, that will be more flexible.